syntax = "proto3";

// https://github.com/protocolbuffers/protobuf/releases
// cd seata/rm/datasource/undo/parser
// protoc --python_out=./proto branch_undolog.proto

package seata.rm.datasource.undo.parser.proto;

import "google/protobuf/any.proto";

message BranchUndoLog {
  optional string xid = 1;
  optional string branch_id = 2;
  repeated SQLUndoLog sql_undo_logs = 3;
}

message SQLUndoLog {
  optional string sql_type = 1;
  optional string table_name = 2;
  optional TableRecords before_image = 3;
  optional TableRecords after_image = 4;
}

message TableRecords {
  optional TableMeta table_meta = 1;
  optional string table_name = 2;
  repeated Row rows = 3;
}

message TableMeta {
  optional string table_name = 1;
  map<string, ColumnMeta> all_columns = 2;
  map<string, IndexMeta> all_indexs = 3;
}

message ColumnMeta {
  optional string table_cat = 1;
  optional string table_schema_name = 2;
  optional string table_name = 3;
  optional string column_name = 4;
  optional int32 data_type = 5;
  optional string data_type_name = 6;
  optional int32 column_size = 7;
  optional int32 decimal_digits = 8;
  optional int32 num_prec_radix = 9;
  optional int32 null_able = 10;
  optional string remarks = 11;
  optional string column_def = 12;
  optional int32 sql_data_type = 13;
  optional int32 sql_datetime_sub = 14;
  optional google.protobuf.Any char_octet_length = 15;
  optional int32 ordinal_position = 16;
  optional string is_nullable = 17;
  optional string is_autoincrement = 18;
}

message IndexMeta {
  repeated ColumnMeta values = 1;
  optional bool non_unique = 2;
  optional string index_qualifier = 3;
  optional string index_name = 4;
  optional int32 type = 5;
  enum IndexType {
    PRIMARY = 0;
    NORMAL = 1;
    UNIQUE = 2;
    FULL_TEXT = 3;
  }
  optional IndexType index_type = 6;
  optional string asc_or_desc = 7;
  optional int32 cardinality = 8;
  optional int32 ordinal_position = 9;
}

message Row {
  repeated Field fields = 1;
}

message Field {
  optional string name = 1;
  enum KeyType {
    NONE = 0; // not exists
    NULL = 1;
    PRIMARY_KEY = 2;
  }
  optional KeyType key_type = 2;
  optional int32 type = 3;
  optional google.protobuf.Any value = 4;
}